src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageClientMiddleLayer.cs

```diff
@@ -0,0 +1,215 @@
+            _documentManager = documentManager;
+            _serviceProvider = serviceProvider;
+        }
+
+        public override bool CanHandle(string methodName)
+        {
+            return methodName == Methods.TextDocumentOnTypeFormattingName;
+        }
+
+        public override Task HandleNotificationAsync(string methodName, JToken methodParam, Func<JToken, Task> sendNotification)
+        {
+            return null;
+        }
+
+        public override async Task<JToken> HandleRequestAsync(string methodName, JToken methodParam, Func<JToken, Task<JToken>> sendRequest)
+        {
+            if (methodName == Methods.TextDocumentOnTypeFormattingName)
+            {
+                var emptyResult = JToken.FromObject(Array.Empty<TextEdit>());
+                var requestParams = methodParam.ToObject<DocumentOnTypeFormattingParams>();
+                if (requestParams.Options.OtherOptions == null)
+                {
+                    requestParams.Options.OtherOptions = new Dictionary<string, object>();
+                }
+
+                requestParams.Options.OtherOptions[LanguageServerConstants.ExpectsCursorPlaceholderKey] = true;
+                var token = JToken.FromObject(requestParams);
+                var result = await sendRequest(token);
+                var edits = result?.ToObject<TextEdit[]>();
+                if (edits == null)
+                {
+                    return emptyResult;
+                }
+
+                await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+                if (!_documentManager.TryGetDocument(requestParams.TextDocument.Uri, out var documentSnapshot))
+                {
+                    return emptyResult;
+                }
+
+                Position cursorPosition = null;
+                var filteredEdits = new List<TextEdit>(edits.Length);
+                foreach (var edit in edits)
+                {
+                    if (edit.NewText.Contains(LanguageServerConstants.CursorPlaceholderString))
+                    {
+                        var newEdit = new TextEdit();
+                        newEdit.Range = edit.Range;
+                        newEdit.NewText = edit.NewText.Replace(LanguageServerConstants.CursorPlaceholderString, string.Empty);
+                        filteredEdits.Add(newEdit);
+
+                        cursorPosition = edit.Range.Start;
+                    }
+                    else
+                    {
+                        filteredEdits.Add(edit);
+                    }
+                }
+
+                ApplyTextEdits(filteredEdits, documentSnapshot.Snapshot, documentSnapshot.Snapshot.TextBuffer);
+
+                if (cursorPosition != null)
+                {
+                    var fullPath = GetLocalFilePath(requestParams.TextDocument.Uri);
+                    VsShellUtilities.OpenDocument(_serviceProvider, fullPath, VSConstants.LOGVIEWID.TextView_guid, out _, out _, out _, out var textView);
```

Comment:
Hmm, I wonder what happens here if you have multiple text views of the same doc

Reply:
This is pretty much what they were doing https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FRemoteLanguage%2FImpl%2FUtilities.cs. I could also use `GetActiveDocument` api and use that text view. What do you think?

Reply:
Ya that'd probably be best

Reply:
Actually I just tested it out with two text views and it works correctly. It updates both.

Reply:
But if you put your cursor in the secondary text view does it properly move the cursor in the correct textview?